PRO-2355 document component admin UI extension patterns

boutell opened the pull request.

boutell commented on docs/guide/custom-ui.md:
```diff
@@ -0,0 +1,62 @@
+# Customizing the Apostrophe user interface
```
I will probably rename this file to narrow its scope to "overriding admin UI Vue components" only. Larger topic is too big for this document although I need to do an overview guide as well which talks about apps vs components and the bus.

boutell commented on docs/guide/custom-ui.md:
```diff
@@ -0,0 +1,62 @@
+
+## Registering custom field types
+
+Apostrophe's [schema field types](content-schema.md) cover many situations, but you might wish to add a new one.
```
I need to move this to another guide for later.

boutell commented on docs/guide/custom-ui.md:
```diff
@@ -0,0 +1,62 @@
+
+Apostrophe's [schema field types](content-schema.md) cover many situations, but you might wish to add a new one.
+
+## Creating custom Vue components
```
I need to move this after the "by name" material because it is the exception to that rule in this particular dev story.

abea commented:
@boutell Are you looking for an initial review?

boutell commented:
This is now ready for proper review. In the end I was able to cover the three major cases right now — override by name, override by configuration (for situations where subclassing makes override by name a bad idea), and custom schema fields, which includes a full example.

You can find a working example of the custom schema field shown here in the `star-rating-field` branch of `testbed`, which I'm also pushing as a PR, seems like a good idea to validate we don't break these.

abea commented on docs/guide/custom-ui.md:
```diff
@@ -0,0 +1,207 @@
+
+### Implementing the server-side part
+
+Any module can register a schema field type on the server side, like this one, which provides a range field with two inputs, one for the low end of the range and one for the high end.
```
"provides" to "creates"

abea replied:
I don't understand the low end/high end part of this. It would also help to include an initial description that this will be a 1-5 rating field to help people contextualize the next parts.

boutell replied:
Oops, that was a leftover.

abea commented on docs/guide/custom-ui.md:
```diff
@@ -0,0 +1,207 @@
+In `addStarRatingFieldType`, we invoke `self.apos.schema.addFieldType` to add our custom field type on the server side. We provide:
+
+* `name`, which can be used as a `type` setting when adding the field to a schema.
+* `convert`, a function to be used to sanitize the input and copy it to a destination. We pass our `convert` method for this purpose. Methods of our module are available as properties of `self`.
```
> We pass our `convert` method for this purpose.

This took me a minute to work out since it's in the description of a `convert` function. Maybe we should name the method something slightly different for clarity.

abea commented on docs/guide/custom-ui.md:
```diff
@@ -0,0 +1,207 @@
+
+* `name`, which can be used as a `type` setting when adding the field to a schema.
+* `convert`, a function to be used to sanitize the input and copy it to a destination. We pass our `convert` method for this purpose. Methods of our module are available as properties of `self`.
+* `component`, the name of a Vue component to be displayed when editing the field.
```
This should be `vueComponent`

abea commented on docs/guide/custom-ui.md:
```diff
@@ -0,0 +1,207 @@
+* `convert`, a function to be used to sanitize the input and copy it to a destination. We pass our `convert` method for this purpose. Methods of our module are available as properties of `self`.
+* `component`, the name of a Vue component to be displayed when editing the field.
+
+In `convert`, we sanitize the input and copy it from `data[field.name]` to `object[field.name]`. Since we must not trust the browser, we take care to sanitize it with [the `launder` module](https://npmjs.com/package/launder), which is always available as `apos.launder`. But we can validate the input any way we like, as long as we never trust the input.
```
Reminder to rename the method to be distinct

abea reviewed (changes requested):
A bunch of line notes and suggestions. The field type tutorial is awesome and deserves to be its own page.

boutell commented on docs/guide/custom-ui.md:
```diff
@@ -0,0 +1,207 @@
+| `@apostrophecms/piece-type`  | `components.managerModal` | `AposDocsManager`  |
+| `@apostrophecms/piece-type`  | `components.editorModal`  | `AposDocEditor`    |
+| `@apostrophecms/page`        | `components.managerModal` | `AposPagesManager` |
+| `@apostrophecms/page`        | `components.editorModal`  | `AposDocEditor`    |
```
Saying `page-type` would result in people configuring that module, which would not work. I will add a note.

boutell commented:
Implemented Bea's notes, added custom column coverage, and reorganized in a better pedagogical order that emphasizes adding things (custom schema field types, custom columns) over overriding things

abea commented:
@boutell The big thing I'm seeing as I start review is that the custom field types section is still part of the same UI page. That should be a separate code recipe page (how to).

abea commented on docs/guide/custom-ui.md:
```diff
@@ -1,39 +1,292 @@
+
+1. Core Apostrophe modules
+2. Installed and project-level modules, in the order they are configured in `app.js`
+that is configured. For instance, if the last module in our project's `app.js` modules list contains a `ui/apos/components/AposLogPadless.vue` file, that logo will be used in the admin bar, in place of the version that is normally loaded from Apostrophe core or in any module configured earlier.
```
There's odd line breaking here. The list is useful for quick scanning, so the "For instance" should be the start of a new paragraph. The "that is configured" can be back with the previous line.

For doc review and diffing it's helpful to avoid extra line breaks.

abea replied:
Actually "that is configured." seems like it should be deleted.

abea commented on docs/guide/custom-ui.md:
```diff
@@ -1,39 +1,292 @@
+
+## Overriding standard Vue.js components through configuration
+
+There can be only one `AposDocsManager` Vue.js component definition in a project, but sometimes we need different behavior for a specific piece type only. We could work around this by overriding a core component and adding conditional logic, but this results in code that is hard to maintain, and also means we are stuck maintaining a copy of a complex component and missing out on bug fixes and imp ...
```
There are a few previous comments that you don't seem to have addressed here.

Remove "Vue.js" from "only one `AposDocsManager` Vue.js component in a project". Not critical, but repetitive.

Remove "only" at the end of the second sentence as it's redundant.

We could work around this by overriding" to "We could override" (then, "and add"). Period after "hard to maintain," then new sentence.

> It would be better if we could...

"Could" isn't necessary since the purpose here is to show they can. Simply "It would be better to specify...".

abea commented on docs/guide/custom-ui.md:
```diff
@@ -1,39 +1,292 @@
+::: warning
+* Altering the UI should be done rarely and carefully. When possible, add new functionality like custom schema field types and custom manager view columns. Avoid overriding components entirely unless absolutely necessary.
+* Overriding a UI component prevents the project from benefiting from future UI improvements and bug fixes related to that component.
+* Make sure there is not a better way to achieve the desired goal. This includes [asking for help in discord](https://chat.apostrophecms.org) and [requesting](https://portal.productboard.com/apostrophecms/1-product-roadmap/tabs/1-under-consideration) or [contributing](https://github.com/apostrophecms/apostrophe/blob/main/CONTRIBUTING.md#apostrophecms-contribution-guide) new features for the core.
```
"Discord" should be capitalized

abea commented on docs/guide/custom-ui.md:
```diff
@@ -0,0 +1,207 @@
+The `AposWidget` component has **nothing to do with a typical site visitor experience.** It is used only when displaying our widget while the page is in edit mode. While overriding this component is rare, but the `@apostrophecms/rich-text-widget` module does so to provide a "click the text to edit" experience for rich text widgets. If you're just trying to enhance your widget with frontend JavaSc ...
+:::
+
+## Registering custom field types
```
This should be a code recipe page instead. It's a separate section in the navigation. If you break this out to a new page I can help with the nav.

I see the value for this topic in terms of expanding UI options. It would make sense to have a paragraph that introduces this idea and links over to the code recipe. There's simply more to custom field types than customizing UI.

abea commented on docs/guide/custom-ui.md:
```diff
@@ -1,39 +1,292 @@
+
+Another common extension is a custom column for the piece type manager. By default the manager modal displays the title, certain flags, and the last time the piece was updated. We can extend this to display our custom star ratings.
+
+### Implementing the server-side part
```
"Server-side configuration" might be more descriptive here. Especially with moving the previous section to its own page, the simpler parallelism might not help as much as being more descriptive.

boutell replied:
It's more than configuration actually (we must write a sanitizer) so I like this language. Moved to the other page though.

abea commented on docs/guide/custom-ui.md:
```diff
@@ -1,39 +1,292 @@
+    add: {
+      rating: {
+        label: 'Rating',
+        component: 'ColumnStarRating'
```
There is no description of what these are. This could be a use case for inline comments with a very brief description of the column properties (optionally using 👇 as I have elsewhere).

abea replied:
I'm specifically talking about the label and component properties.

boutell commented:
I've refactored as requested. It makes sense for it to be a separate article. However I wanted to run a concern by you:

I'm not convinced this new article belongs in "cookbook" and not "guide." Compare it to "custom batch operations" or "custom widgets." It's similar in structure and its role is similarly to document a clearly defined core feature with examples. "Cookbook" so far seems to be the domain of things that aren't Apostrophe features *per se* but Apostrophe devs want to know more about, like navigation or specific hosting situations. Custom schema field types aren't a special case or a particular application, they are a primary feature with a well-defined boundary. As a user I would expect to be able to find them in the side nav of the guide with similar topics, but instead they are in the cookbook which is not part of that. "Advanced topics" within the guide seems like the more natural place.

abea commented:
That's a good point. It makes sense to me to put it in a nav section such as "Other customizations" parallel to "Other module config."

boutell commented:
Would you mind doing that part after this PR lands since you have the
clearest sense of how to handle the nav?


On Wed, Jan 5, 2022 at 10:39 AM Alex Bea ***@***.***> wrote:

> That's a good point. It makes sense to me to put it in a nav section such
> as "Other customizations" parallel to "Other module config."
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/apostrophecms/a3-docs/pull/153#issuecomment-1005833487>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAH27P4U6HVGCFKNT6M6H3UURQ2XANCNFSM5IW4IABA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 

THOMAS BOUTELL | CHIEF TECHNOLOGY OFFICER
APOSTROPHECMS | apostrophecms.com | he/him/his

abea commented:
sure
